### ISSUE TITLE
Add Authorization to reference implementation

### ISSUE THREAD

**Issue by R2wenD2** [grafeas/grafeas#28]

**Comment by hpejustin** [comment]
@R2wenD2 RBAC will be a good choice.

**Comment by ddimtirov** [comment]
Add an option to serve 404 rather than 401 - in some cases, the mere knowledge that a resource exists is significant in security context.

**Comment by vtsao** [comment]
First we'll need to define the permissions/roles we have in the API layer for v1beta1. Currently we have the following:

"notes.get"
"notes.list"
"notes.create"
"notes.update"
"notes.delete"

"occurrences.get"
"occurrences.list"
"occurrences.create"
"occurrences.update"
"occurrences.delete"

"notes.listOccurrences"
"notes.attachOccurrence"

"operations.get"
"operations.list"
"operations.create"
"operations.update"
"operations.delete"

I think @furuholm was interested in these. Let me know if anybody has thoughts on this. Otherwise we'll go with these for now.

**Comment by R2wenD2** [comment]
We'll also need projects roles @vtsao 

Propose:

projects.create
projects.get
projects.list
projects.update

**Comment by vtsao** [comment]
Ah yeah missed that.

We don't allow updating of projects right now since a project just has a name that can't be changed. We have delete though.

So:
projects.update -> projects.delete

**Comment by furuholm** [comment]
I want to preface this by stating that I am not an expert in this area so please take this with a grain of salt. @R2wenD2 I'm thinking you probably have one or two subject experts at google that might be able to provide some feeback :). Also, there probably are libraries (or services?) out there that can do a lot of the heavy lifting for us. Would be great to get some suggestions!

Back to the issue: 
The way I see this what you have provided is more a list of resources than roles, but that is ok, we just need some more stuff if we are going the RBAC route.

We have been discussing OAuth for authentication, which I would say in reality means OpenID Connect (OIDC). This is what is used in Kubernetes when people say that they use OAuth for authentication in Kubernetes. This should give us support for e.g. CoreOS dex and Google. 

The way I understand it there needs to be a way to manange users if we chose OIDC. OIDC targets authentication i.e. gives us who the user is (or more exactly the users email address) but it does not solve authorization (what is the user allowed to do). So given a user we need to determine what resources this user is allowed to access. If we go with RBAC this is managed through roles that are assigned to users (role binding). E.g. a user assigned the admin role might be allowed to access all resources. This means that we need some kind of user registry that contains users and their assigned roles. It probably also means a few new resources (get, list, create, delete and modify user).

To summarize, if we chose the route I have outlined above we need to support:
- OIDC (for authentication)
- Resources (the list you provided)
- Users
- Roles and role bindings

A way to make this simpler could be to make the resources/roles less flexible by "hard coding role bindings", i.e. identify a number of user types and grant them access to some set of resources. I would suggest this simplification if possible to keep the complexity down.

**Comment by furuholm** [comment]
On the Grafeas status VC a couple of weeks ago I promised @skelterjohn I would write something up about my thoughts on this topic. So here it goes...

After having implemented auth for a another golang app I have put some more thought on this. Below is a rough outline of a design. Would like to get the time to do this some time during the next couple of months or so but I can't promise anything unfortunately. Still I think it could be valuable to get the discussion going.

As discussed above I suggest RBAC for controlling access to resources. RBAC in this context would mean that we define resources (e.g. projects or notes) and verbs (e.g. read, create, edit, delete). Roles are  defined in terms of allowed verbs and resources. Users are assigned roles. Once this is in place we can conclude e.g. that user A is allowed to read notes in project P, but not allowed to create occurences in the same project.

There are [some](https://github.com/casbin/casbin) [libraries](https://github.com/mikespook/gorbac) out there that could be useful, but I would suggest implementing the RBAC part ourselves. It's not that much code (it's basically matching users against rules that are attached to each role). We can take a lot of inspiration from the Kubernetes RBAC implementation, e.g. users and roles could be similar to corresponding Kubernetes concepts and I think coneptually Kubernetes namespaces maps pretty good to Grafeas projects (when it comes to the RBAC implementation that is). Some Kubernetes RBAC concepts could probably be skipped, e.g. cluster role binding provides a level of flexibility that I don't think we need. To start with I suggest we set up a number of predefined roles, e.g. admin, project/notes/occurrence reader and writer (namespaced per project). Not sure if user groups are necessary. Maybe we can start without them and add them if/when needed?

For authentication I suggest we use OAuth2/OIDC. The nice thing with OIDC is that Grafeas does not have to store any passwords, instead authentication is handled by an external resource that allows us to establish the identity of the user (using a JSON web token provided by the user). Managing users is then a matter of keeping a list of users and their respective roles. I have previously used [github.com/coreos/go-oidc](https://github.com/coreos/go-oidc) for verifying JSON web tokens in go backends and suggest we use that here as well. BTW: [Dex](https://github.com/dexidp/dex) can map to different backends, including LDAP which may be of interest for on-prem solutions.

As mentioned above there needs to be a way to manange users and assign them to roles. I would suggest adding a user registry to the database with a corresponding API. As discussed on the meeting this should probably not be part of the "official Grafeas API". For those that have external user registries it would be nice to support e.g. google IAM to manage users and roles as an alternative to the built in (database) user registry. I'm thinking that would have to be done later but that we keep this future extension point in mind when implementing the user registry. To bootstrap the user registry we could use a config parameter that specifies the email address of the initial admin?

For web clients (out of scope for this issue) there are javascript libraries (e.g. [oidc-client](https://github.com/IdentityModel/oidc-client-js)) that can be used for managing the login procedure. For CLI clients maybe something like the login process in the gcloud CLI could be done?

Rough todo-list: 

- [ ] Add RBAC rules and user mapping
- [ ] Add OAuth JWT verification/decoding (using [go-oidc](https://github.com/coreos/go-oidc))
- [ ] Add resource/verb verification to all API entry points
- [ ] Add user registry to data store
- [ ] Add user manangement API
- [ ] Add user bootstrapping
- [ ] Add ability to config auth on or off

Thoughts?

**Comment by vtsao** [comment]
Thanks @furuholm, using RBAC for authorization and OAuth2/OIDC for authentication sounds good. We'll have to decide how to prioritize this. Adding @aysylu since she might be the one working on this.

**Comment by bharathrajbk** [comment]
I am keen to see this feature being implemented as well! :+1:
